---
title: "Nrf2 BL6 PEITC 16S Microbiome Logit-Transformed Data Modeling"
output:
  html_notebook: default
  html_document:
    df_print: paged
  pdf_document: default
---


```{r Data, warning=FALSE,echo=FALSE,message=FALSE}
require(data.table)
require(phyloseq)
require(ggplot2)
require(plotly)
require(DT)
source("source/functions_v2.R")

# Load data----
# Counts
load("data/ps.RData")

# Taxonomy
load("data/taxa.plus.RData")
taxa <- data.table(seq16s = rownames(taxa.plus),
                   taxa.plus)

# Samples
# ps@sam_data
load("data/samples.RData")
samples$Sample <- substr(x = samples$Name,
                         start = 1,
                         stop = 5)
samples$Sample[samples$Sample %in% c("4A_S1",
                                     "4B_S2",
                                     "4C_S3")] <- 
  substr(x = samples$Sample[samples$Sample %in% c("4A_S1",
                                                  "4B_S2",
                                                  "4C_S3")],
         start = 1,
         stop = 2)
DT::datatable(samples)
```

### Taxonomy Ranks:
#### **K**ing **P**hillip **C**an n**O**t **F**ind **G**reen **S**ocks
* Kingdom                
* Phylum                    
* Class                   
* Order                   
* Family     
* Genus     
* Species  


## Relative Abundance in Samples at Different Taxonomic Ranks
```{r Tax, warning=FALSE,echo=FALSE,message=FALSE}
# keep controls and PEITCs at weeks 5 and 9 only
w59 <- prune_samples(samples = sample_names(ps)[!(sample_names(ps) %in%
                                                    c("4A",
                                                      "4B",
                                                      "4C",
                                                      "Undetermined"))],
                    x = ps)
# OTU table
otu <- t(w59@otu_table@.Data)
otu <- data.table(seq16s = rownames(otu),
                  otu)

# Merge taxonomy and counts tables----
dt1 <- merge(taxa,
             otu,
             by = "seq16s")
dt1$seq16s <- NULL

# Remove archea and eucaryota----
dt1 <- droplevels(dt1[Kingdom == "Bacteria", ])
```

### 1. Class
```{r Class, warning=FALSE,echo=FALSE,message=FALSE,fig.width=10,fig.height=7}
tax <- "Class"
dt2 <- counts_ra_by_tax_rank(dt1, tax)

mu <- aggregate(dt2$RA,
                by = list(Tax = dt2$Tax,
                          Week = dt2$Week,
                          Diet = dt2$Diet),
                FUN = "mean")

lvls <- aggregate(dt2$RA,
                  by = list(Tax = dt2$Tax),
                  FUN = "mean")
lvls <- lvls$Tax[order(lvls$x,
                       decreasing = TRUE)]
lvls
dt2$Tax <- factor(dt2$Tax,
                  levels = lvls)

tmp <- droplevels(dt2[Tax == "Clostridia", ])
m1 <- lm(lgt ~ Diet,
         data = tmp,
         weights = 1/(Counts + 0.5))
summary(m1)

exp(m1$coefficients[2])/(1 + exp(m1$coefficients[2]))

# NO GOOD!
# # All classes
# m2 <- lm(lgt ~ 0 + Diet + Tax,
#          data = dt2,
#          weights = 1/(Counts + 0.5))
# summary(m2)
# anova(m2)
```

### 2. Order
```{r Order, warning=FALSE,echo=FALSE,message=FALSE,fig.width=10,fig.height=7}

```

### 3. Family
```{r Family, warning=FALSE,echo=FALSE,message=FALSE,fig.width=10,fig.height=7}

```

### 4. Genus
```{r Genus, warning=FALSE,echo=FALSE,message=FALSE,fig.width=10,fig.height=7}

```